Adding tests for org.openmrs.util.Format in org.openmrs.util.FormatTe…

winglam opened the pull request:
…st. The added tests calls org.openmrs.util.Format.format with nulls and valid arguments. The following tests fail because it is expecting org.openmrs.util.Format.format to return null. format_shouldAllowAllNullParameters, format_shouldAllowLocaleTypeNullParameters, format_shouldAllowLocaleNullParameters, and format_shouldAllowDateLocaleNullParameters gets a NullPointerException instead and format_shouldReturnDateWithNullType and format_shouldAllowDateNullParameters receives an empty string as the return.

A bug report related to this pull request can be found on: 
https://issues.openmrs.org/browse/TRUNK-4849

dkayiwa commented:
@winglam did you take a look at this? https://wiki.openmrs.org/display/docs/Unit+Testing+With+at-should+Annotation

winglam commented:
@dkayiwa Thanks for pointing out the @should annotation. I have added that, other documentation, null checks to org.openmrs.util.Format and changed the tests in org.openmrs.util.FormatTest to respect the documentation.

dkayiwa commented:
Can you squash these commits into one? https://wiki.openmrs.org/display/docs/Pull+Request+Tips

winglam commented:
I have squashed the commits.

dkayiwa commented on api/src/main/java/org/openmrs/util/Format.java:
```diff
@@ -62,8 +62,25 @@ public static String format(Date date) {
+	 * @return empty string if one of the parameters is null. Otherwise a string object for the date such that it is
+	 * formatted according to locale and the amount of information it contains is determined by type.
+	 * @should not fail with null date
+	 * @should not fail with null locale
```
Did you add a should annotation for each and every test you added?

winglam replied:
No I did not. However I believe these three annotations are sufficient for the 8 tests that I have added. 3 of the 8 tests test the case when one of these parameters are null. Another 3 of the 8 tests test the case when two of the parameters are null. The remaining two tests test the case when all of them are null or none of them are null. The should annotations I have added implies that regardless of whether any of the parameters are null or not, the method should not fail.

dkayiwa replied:
If you want to break the convention, you will need to start a discussion on talk to defend your argument.
We even have a tool that generates a test for each should annotation. https://wiki.openmrs.org/display/docs/Generate+Test+Case+Plugin

winglam replied:
I did not realize that each should annotation needs to correspond to a test. I have added/changed the should annotations so that each one corresponds to a test case now. Thanks.

dkayiwa commented:
@winglam if you check again on the above wiki page, you will notice that the should annotation matches the method name.

winglam commented:
Sorry, I seem to have overlooked that from before. I have now changed the method names to match the should annotation.

dkayiwa commented:
@winglam hopefully you are aware of this plugin which should automatically generate test methods from the should annotations: https://wiki.openmrs.org/display/docs/Generate+Test+Case+Plugin
